torch/ao/quantization/fx/_lower_to_native_backend.py

```diff
@@ -130,6 +139,86 @@ def _lower_weighted_ref_module(model: QuantizedGraphModule) -> QuantizedGraphMod
+            pattern = (torch.quantize_per_tensor, func_pattern, MatchAllNode, MatchAllNode, MatchAllNode)
+
+            # Iterate through nodes in the graph to find a match
+            # If there is a match, replace the above pattern with the corresponding quantized op
+            modules = dict(model.named_modules(remove_duplicate=False))
+            nodes = list(model.graph.nodes)
+            for n in model.graph.nodes:
+                if not is_match(modules, n, pattern):
+                    continue
+                q_node = n
+                (func_node, output_scale_node, output_zp_node, dtype) = q_node.args
+                if is_relu:
+                    relu_node = func_node
+                    func_node = relu_node.args[0]
+                else:
+                    relu_node = None
+                input_dq_node = func_node.args[0]
+                weight_dq_node = func_node.args[1]
+
+                # Step 1: Replace quantized weights with packed weights
+                quantized_weight = weight_dq_node.args[0]
+                weight_dtype = quantized_weight.args[4]
+                if has_bias:
+                    bias = func_node.args[2]
+                else:
+                    bias = func_node.kwargs.get("bias", None)
+                prepack_args = (quantized_weight, bias)
+                if ref_func == F.linear:
+                    prepack_op = get_linear_prepack_op_for_dtype(weight_dtype)
+                else:
+                    raise ValueError("Lowering for functional currently only supports linear op")
+                insert_prepack_after = bias if has_bias else quantized_weight
+                with model.graph.inserting_after(insert_prepack_after):
+                    packed_weight = model.graph.create_node("call_function", prepack_op, prepack_args, {})
+
+                # Step 2: Replace reference pattern with the corresponding quantized op
+                with model.graph.inserting_after(output_zp_node):
+                    args = (input_dq_node.args[0], packed_weight, output_scale_node, output_zp_node)
+                    new_func = q_relu_func if is_relu else q_func
+                    new_func_node = create_node_from_old_node_preserve_meta(
+                        model.graph,
+                        ("call_function", new_func, args, {}),
+                        func_node)
+                    q_node.replace_all_uses_with(new_func_node)
+
+                # Clean up: Remove dequantize and quantize nodes and the old func node
+                for dqn in [input_dq_node, weight_dq_node]:
+                    dqn_input = dqn.args[0]
+                    dqn.replace_all_uses_with(dqn_input)
+                    model.graph.erase_node(dqn)
+                model.graph.erase_node(q_node)
+                if is_relu:
+                    model.graph.erase_node(relu_node)
+                model.graph.erase_node(func_node)
+
+                # Step 3: TODO(andrew) Fold weights
```

Comment:
yeah we just need to swap the order of weight folding and lowering: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/convert.py#L648, also we can move fold_weight to _lower_to_native_backend

Reply:
can we verify that folding happens? if all tests passes now we probably haven't do that, maybe we can add a test for it

Reply:
Ok, I added a check at the end of `test_functional_linear`